service/models/wishlist.py

```diff
@@ -11,24 +11,141 @@
+    db.session.commit()
+  
+  def read(self):
+    wishlist = Wishlist.find_by_id(self.id)
+    if wishlist is None:
+      raise EntityNotFoundError("Cannot find wishlist with id %s" % self.id)
+    else:
+      product_ids = [wp.product_id for wp in WishlistProduct.find_all_by_wishlist_id(self.id)]
+      products = []
+      if len(product_ids) != 0:
+        products = Product.find_all_by_id(product_ids)
+      return WishlistVo(self,products).serialize()
   
-  def add_items():
-    pass
+  def add_items(self,product_ids:list):
+    cnt = 0
+    for pid in product_ids:
+      if WishlistProduct.find_by_wishlist_id_and_product_id(self.id,pid) is None:
+        cnt += 1
+        WishlistProduct(wishlist_id=self.id,product_id=pid).create()
+    return cnt
 
-  def delete_items():
-    pass
+  def delete_items(self, product_ids:list):
+    cnt = 0
+    for pid in product_ids:
+      entity = WishlistProduct.find_by_wishlist_id_and_product_id(self.id, pid)
+      if entity is not None:
+        cnt += 1
+        entity.delete()
+    return cnt
+  
+  def delete(self):
+    if self.id is None:
+      logger.info("Wishlist: delete a wishlist with id None")
+      return 0
+    wishlist = Wishlist.find_by_id(self.id)
+    if wishlist is None:
+      logger.info("Wishlist: cannot find this wishlist")
+      return 0
+    else:
+      WishlistProduct.delete_all_by_wishlist_id(self.id)
+      logger.info("Deleting wishlist % s" % self.name)
+      db.session.delete(self)
+      db.session.commit()
+      return 1
+
+  @classmethod
+  def init_db(cls, app:Flask):
+    logger.info("Wishlist: Initializing database")
+    cls.app = app
 
-  def read():
-    pass
+    db.init_app(app)
+    app.app_context().push()
+    db.create_all()
   
-  def delete():
-    pass
+  @classmethod
+  def find_all(cls):
+    logger.info("Wishlist: processing lookup for all wishlists")
+    return cls.query.all()
 
-  def serialize():
-    pass
+  @classmethod
+  def find_by_id(cls, wishlist_id:int):
+    logger.info("Wishlist: processing deletion for id %s ..." % wishlist_id)
+    return cls.query.get(wishlist_id)
+  
+  @classmethod
+  def find_all_by_user_id(cls,user_id:int)->list:
+    logger.info("Wishlist: porcessing lookup for user id: %s ..." % user_id)
+    query_res = cls.query.filter(cls.user_id == user_id)
+    if query_res.count() == 0:
+      return []
+    else:
+      wishlists = [r for r in query_res]
+      res = []
+      for w in wishlists:
+        wishlist_products = WishlistProduct.find_all_by_wishlist_id(w.id)
+        products = []
+        if len(wishlist_products) != 0:
+          products = Product.find_all_by_id([wp.product_id for wp in wishlist_products])
+        res.append(WishlistVo(w,products))
+      
+      return [vo.serialize() for vo in res]
 
-  def deserialize():
-    pass
\ No newline at end of file
+class WishlistVo:
+    def __init__(self,wishlist:Wishlist,products:list) -> None:
```

Comment:
Sorry, I want to make sure I understand what this class does. Is this just a wrapper on top of `Wishlist` that adds `products`? Why not just put it all in one class.

Reply:
This is more of a Java thing. I will try to rewrite this part.

Reply:
I will leave it as it was because of the following reasons:

1. Each model defined here will represent a perfect match between the relations inside database and the real instance.

2. When it comes to testing, having wishlist like this will make testing easier because we can make use of factory to generate instances.